jujugui/static/gui/src/test/test_utils.js

```diff
@@ -120,6 +120,202 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+        return {
+          set: function(key, value) {
+            _attrs[key] = value;
+          },
+
+          get: function(key) {
+            return _attrs[key];
+          }
+        };
+      };
+      var wordpressProvides = {
+        website: {
+          'interface': 'http',
+          scope: 'global'
+        }
+      };
+      var wordpressRequires = {
+        cache: {
+          'interface': 'memcache',
+          scope: 'global'
+        },
+        db: {
+          'interface': 'mysql',
+          scope: 'global'
+        },
+        nfs: {
+          'interface': 'mount',
+          scope: 'global'
+        }
+      };
+      var mysqlProvides = {
+        db: {
+          'interface': 'mysql',
+          scope: 'global'
+        },
+        'db-admin': {
+          'interface': 'mysql-root',
+          scope: 'global'
+        },
+        'local-monitors': {
+          'interface': 'local-monitors',
+          scope: 'container'
+        },
+        master: {
+          'interface': 'mysql-oneway-replication',
+          scope: 'global'
+        },
+        monitors: {
+          'interface': 'monitors',
+          scope: 'global'
+        },
+        munin: {
+          'interface': 'munin-node',
+          scope: 'global'
+        },
+        'shared-db': {
+          'interface': 'mysql-shared',
+          scope: 'global'
+        }
+      };
+      var mysqlRequires = {
+        ceph: {
+          'interface': 'ceph-client',
+          scope: 'global'
+        },
+        ha: {
+          'interface': 'hacluster',
+          scope: 'global'
+        },
+        slave: {
+          'interface': 'mysql-oneway-replication',
+          scope: 'global'
+        }
+      };
+      endpoints = [
+        {
+          charm: MockCharm({
+            provides: wordpressProvides,
+            requires: wordpressRequires
+          }),
+          name: 'wordpress',
+          service: {},
+          type: 'db'
+        },
+        {
+          charm: MockCharm({
+            provides: mysqlProvides,
+            requires: mysqlRequires
+          }),
+          name: 'mysql',
+          service: {},
+          type: 'db'
+        }
+      ];
+    });
+
+    it('finds endpoint matches for relations', function() {
+      var actual = utils.findEndpointMatch(endpoints);
+      var expected = {
+        'interface': 'mysql',
+        scope: 'global',
+        provides: endpoints[1],
+        requires: endpoints[0],
+        provideType: 'db',
+        requireType: 'db'
+      };
+      assert.deepEqual(actual, expected);
+    });
+
+    it('errors when no endpoint matches exist', function() {
+      endpoints[1].charm.set('provides', {});
+      var actual = utils.findEndpointMatch(endpoints);
+      assert.equal(actual.error, 'Specified relation is unavailable.');
+    });
+
+    it('errors when multiple explicit endpoint matches exist', function() {
+      // In order to force an ambiguous relation, we need to null these out.
```

Comment:
It's best to actually supply extra data so that the relation is truly ambiguous. This is relying on the internal implementation of the method in order to short circuit a case. So if this method ever changes this test won't properly represent the output.